### PR TITLE
Add LTS version of GnuPG and update libgcrypt to latest LTS version

### PR DESCRIPTION
**Pull request by afh**
###### Description of changes

**In Brief:**

* Add package to install GnuPG 2.2.41 and libgcrypt 1.8.10 both listed as the official Long Term Support (LTS) versions on the GnuPG project's [download page](https://www.gnupg.org/download/index.html)
* Add `gnupg-lts` and `libgcrypt-tls` package names
* Make `gnupg-lts` depend on `libgcrypt-tls`
* Refactor shared code between 22.nix and 23.nix into common.nix

ℹ️ Updating libgcrypt from 1.5.x to 1.8.x seemed okay to me as I did not quickly find any packages depending on it.

**In Detail:**

The latest change of the GnuPG package (see PR #207071) has been met with both acclaim and some uncertainty.
Folks were uncertain whether NixOS should only provide the latest version or additionally offer an LTS version. The GnuPG 2.2.27 package was was replaced by the GnuPG 2.3.3 package (see 1ee8f771320b45651d21e4cdbf15141e1c143887).

A quick investigation shows that several distributions chose the GnuPG LTS version as the default (see [my comment on #207071](https://github.com/NixOS/nixpkgs/pull/207071#issuecomment-1382567197)).

[@mweinelt's comment](https://github.com/NixOS/nixpkgs/pull/207071#issuecomment-1420051633) mentioning _"some interoperability concerns voiced on the [distributions@](https://lore.kernel.org/distributions/F30D6590-3E0C-4865-A944-7DE118A619CF@gentoo.org/T/#u) list, where gnugp with 2.4.0 is incompatible with openpgp implementations following the proposed(?) IETF standard [see [draft-ietf-openpgp-crypto-refresh](https://datatracker.ietf.org/doc/draft-ietf-openpgp-crypto-refresh/)]"_ inspired me to try and (re-)introduce an LTS version of GnuPG and libgcrypt, the result of which is this PR.

✅ ~~**👋 Call for help:**~~

~~Folks familiar with both or either of the packages, please chime in with additional context that may have been lost and whether the patches that are commented out can be removed or need to be updated.~~

~~Until the above call for help has been resolved I believe this PR should continue to be a draft.~~

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by peti — APPROVED** [review]
I am for it. I would use the latest version by default, but offer the LTS version for those who want it, just like this PR does. 👍

**Comment by afh** [comment]
Thanks for your comment, @peti. Delighted to hear you are for it 🙂 

Would you happen to have any additional context you could share on the state of the patches as mentioned in the **👋 Call for help** [above](https://github.com/NixOS/nixpkgs/pull/215109#issue-1574185783) by any chance?

**Comment by peti** [comment]
>Would you happen to have any additional context you could share on the state of the patches as mentioned in the **👋 Call for help** [above](https://github.com/NixOS/nixpkgs/pull/215109#issue-1574185783) by any chance?

No, I'm afraid not.

**Comment by doronbehar** [comment]
The `common.nix` file was not pushed.

**Comment by afh** [comment]
Thanks for bringing this to my attention, @doronbehar. It seems I'm suffering from a severe case of Monday-morning 🧠 today 😅.

**Comment by afh** [comment]
Friendly ping to @lsix, @vrthra, @fpletz about whether you may have more context about the [commented tests](ttps://github.com/NixOS/nixpkgs/pull/215109/files#diff-c0bb3e8e8b06eb65026e7344e056ac3509ce758d8f9f61fa451d04287668700eR13) and whether they can be removed or need to be updated.

**Comment by afh** [comment]
@trofi and @winterqt you've been so very helpful in moving forward with other PRs (#194099, #215699), mind lending a hand here to get the right people involved to get :eyes: on this PR?

The main question to resolve is what should happen with the [patches](https://github.com/NixOS/nixpkgs/pull/215109/files#diff-c0bb3e8e8b06eb65026e7344e056ac3509ce758d8f9f61fa451d04287668700eR15-R27) that do not apply (cleanly)? Once that is resolved I feel the PR is ready for review.

**Comment by trofi** [comment]
> @trofi and @winterqt you've been so very helpful in moving forward with other PRs (#194099, #215699), mind lending a hand here to get the right people involved to get eyes on this PR?
> 
> The main question to resolve is what should happen with the [patches](https://github.com/NixOS/nixpkgs/pull/215109/files#diff-c0bb3e8e8b06eb65026e7344e056ac3509ce758d8f9f61fa451d04287668700eR15-R27) that do not apply (cleanly)? Once that is resolved I feel the PR is ready for review.

I am not familiar with `gnupg`, thus I can only provide general review guidance that might not be in line with maintaners' preferences:

1. Avoid using `common.nix` and just inline the contents into all versions you provide.
2. Check if patches that failed to apply are already upstreamed. Some of them look like backports. But you are introducing older versions. Who nows what we need to backport compared to latest one.
3. Avoid the suffixes like `libgcrypt-lts`. Using versioned ones like `libgcrypt_1_8` where needed should be enough.

Generally if there are no known problems with latest version and there are no packages that rely on outdated versions I would suggest not to provide outdated versions. Especially for security packages. Upstream should know better when to release.

**Comment by afh** [comment]
Thanks for your helpful comments, @trofi, and apologies for taking so long to reply.
I've incorporated your comments into this PR and will look into whether the patches that do not apply have already been merged upstream…

**Comment by trofi** [comment]
> error: undefined variable 'pcsclite'

Can you have a look at `ofborg-eval` presubmit step on why it fails?

**Comment by afh** [comment]
Of course! When reverting the common.nix change I seem to have removed pcsclite package from the list of declared dependencies. This has been corrected and I'll look into the patches in the coming days.

**Comment by afh** [comment]
Finally got around to inspect the patches, @trofi.

Obsoleted patches have been removed, other patches have been updated to apply cleanly.

I'm happy to remove the `0001-dirmngr-Only-use-SKS-pool-CA-for-SKS-pool.patch` altogether, if folks prefer to not have it lie around unused.
The thing with this patch is that the code it patches is commented out upstream with the note that it has been _"[d]isabled […] to due problems with the standard hkps pool."_

**Comment by doronbehar** [comment]
> I'm happy to remove the `0001-dirmngr-Only-use-SKS-pool-CA-for-SKS-pool.patch` altogether

That sounds great! Could this removal be applied for the non LTS Gnupg?

**Comment by afh** [comment]
Yes, it can, @doronbehar, and I pushed an appropriate change for the removal.

For reference here are links to the upstream source showing the comment and commented code:
* [GnuPG 2.4](https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=blob;f=dirmngr/http.c;h=20f71f61bf9f59f259710506ff0557af9452fb37;hb=9f27e448bf1f825906f3c53e3428087d34bbd8fc#l798)
* [GnuPG 2.2](https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=blob;f=dirmngr/http.c;h=946234ea6a430b5e16ee4e736c61e2398786bbb8;hb=da044776311e2951e1deea696939346123bde029#l769)

**Review by doronbehar — COMMENTED** [review]
Maybe we should rename the files in both `gnupg/` and in `libgcrypt/` as `default.nix` and `lts.nix`?

Besides that, looks good to me and I approve and encourage the changes, especially the cleanup commits (libgcrypt update & patch cleanup).

**Comment by afh** [comment]
Thanks for the review, @doronbehar. Glad to hear you approve and encourage the suggested changes. 🙂 

@trofi [suggested](https://github.com/NixOS/nixpkgs/pull/215109#issuecomment-1444286317) to _"[a]void the suffixes like libgcrypt-lts. Using versioned ones like libgcrypt_1_8 where needed should be enough."_ Trofi's comment was about directed at package names; not sure how much of a difference that makes regarding the package filenames.

**Comment by doronbehar** [comment]
I think this is a matter of taste and the maintainers (who unfortunately don't share their voice here) should decide on those details.

**Comment by peti** [comment]
> I think this is a matter of taste and the maintainers (who unfortunately don't share their voice here) should decide on those details.

I don't feel strongly about the filenames. Using the packages version number in the filename seems to be common practice, but the current scheme works fine, too. No big deal. I'm fine with the PR as it is.

**Comment by afh** [comment]
Thanks for chiming in @peti.

Kindly requesting @fpletz and @vrthra to share their voice on the proposed changes and opinion regarding nix filename and package naming 🙂

**Comment by afh** [comment]
@doronbehar, @trofi, @peti the maintainers have been unresponsive here, do any of you have merge rights to possibly get this in for the 23.05 release?

**Review by trofi — APPROVED** [review]

**Comment by trofi** [comment]
Oh, isn't 5000+ rebuilds a bit too much for `master`?

**Comment by K900** [comment]
32000 rebuilds, no less. Backing out.

File: pkgs/tools/security/gnupg/22.nix
Comment (SuperSandro2000) [inline pkgs/tools/security/gnupg/22.nix:10]: with lib should not be over the entire file

**Comment by afh** [comment]
Thanks everyone for chiming in. I'll work on the requested changes and will issue a new PR with `staging` as its merge base. Stay tuned!

**Pull request merged**
